File: app/src/main/java/com/elitec/alejotaller/feature/product/domain/caseUse/RefreshProductsByIdsCaseUse.kt
package com.elitec.alejotaller.feature.product.domain.caseUse

import android.util.Log
import com.elitec.alejotaller.feature.product.domain.entity.Product
import com.elitec.alejotaller.feature.product.domain.repository.ProductRepository

/**
 * Refresh selectivo offline-first tras stock:changed (Pusher).
 * Solo los productIds de la señal se leen de Appwrite y actualizan cache local.
 */
class RefreshProductsByIdsCaseUse(
    private val repository: ProductRepository
) {
    suspend operator fun invoke(productIds: List<String>): Result<List<Product>> = runCatching {
        val ids = productIds.filter { it.isNotBlank() }.distinct()
        val updated = mutableListOf<Product>()
        for (id in ids) {
            val product = repository.refreshFromRemote(id)
            if (product != null) updated += product
        }
        Log.i(TAG, "event=refresh_by_ids count=${updated.size}/${ids.size}")
        updated
    }

    companion object {
        private const val TAG = "RefreshProductsByIds"
    }
}

File: web/src/core/feature/product/domain/caseuse/RefreshProductsByIdsCaseUse.ts
import type { Product } from "../entity/Product";
import type { ProductRepository } from "../repository/product.repository";

/**
 * Refresco selectivo offline-first: solo los productIds señalados por pulse/stock.
 * Fuente de verdad Appwrite → cache Dexie vía repository.getById.
 */
export class RefreshProductsByIdsCaseUse {
    constructor(private readonly repository: ProductRepository) {}

    async execute(productIds: string[]): Promise<Product[]> {
        const ids = [...new Set(productIds.filter(Boolean))];
        const updated: Product[] = [];

        console.info(`[stock-rt] refreshByIds start count=${ids.length} ids=${ids.join(",")}`);

        for (const id of ids) {
            try {
                const product = await this.repository.getById(id);
                if (product) {
                    updated.push(product);
                    console.info(
                        `[stock-rt] refreshByIds OK id=${id} existence=${product.existence} ` +
                            `reserved=${product.reserved ?? 0} available=${product.existence - (product.reserved ?? 0)}`
                    );
                } else {
                    console.warn(`[stock-rt] refreshByIds null id=${id}`);
                }
            } catch (e) {
                console.warn(
                    `[stock-rt] refreshByIds FAIL id=${id}: ${e instanceof Error ? e.message : String(e)}`
                );
            }
        }

        console.info(
            `[stock-rt] refreshByIds done refreshed=${updated.length}/${ids.length}`
        );

        return updated;
    }
}
